src/GoogleFitOnFhir/Services/GoogleFitImportService.cs

```diff
@@ -0,0 +1,119 @@
+            ITelemetryLogger telemetryLogger)
+            : base(options, options?.ParallelTaskOptions?.MaxConcurrency ?? 1)
+        {
+            _googleFitClient = EnsureArg.IsNotNull(googleFitClient, nameof(googleFitClient));
+            _eventHubProducerClient = EnsureArg.IsNotNull(eventHubProducerClient, nameof(eventHubProducerClient));
+            _logger = EnsureArg.IsNotNull(logger, nameof(logger));
+            _telemetryLogger = EnsureArg.IsNotNull(telemetryLogger, nameof(telemetryLogger));
+        }
+
+        public async Task ProcessDatasetRequests(
+            User user,
+            IEnumerable<DataSource> dataSources,
+            string datasetId,
+            AuthTokensResponse tokensResponse,
+            CancellationToken cancellationToken)
+        {
+            var workItems = dataSources.Select(
+                dataSource => new Func<Task>(
+                async () =>
+                {
+                    var dataStreamId = dataSource.DataStreamId;
+                    string pageToken = null;
+
+                    try
+                    {
+                        // Make the Dataset requests, requesting the next page of data if necessary
+                        do
+                        {
+                            _logger.LogInformation("Query Dataset: {0}", dataStreamId);
+                            var medTechDataset = await _googleFitClient.DatasetRequest(
+                                tokensResponse.AccessToken,
+                                dataSource,
+                                datasetId,
+                                cancellationToken,
+                                pageToken);
+
+                            if (medTechDataset == null)
+                            {
+                                _logger.LogInformation("No Dataset for: {0}", dataStreamId);
+                                continue;
+                            }
+
+                            // Save the NextPageToken
+                            pageToken = medTechDataset.GetDataset().NextPageToken;
+
+                            _logger.LogInformation("Create Eventhub Batch");
+
+                            // Create a batch of events for IoMT eventhub
+                            using var eventBatch = await _eventHubProducerClient.CreateBatchAsync(cancellationToken);
+
+                            _logger.LogInformation("Push Dataset: {0}", dataStreamId);
+
+                            // Push dataset to IoMT connector
+                            if (!eventBatch.TryAdd(medTechDataset.ToEventData(user.Id)))
+                            {
+                                throw new Exception("Event is too large for the batch and cannot be sent.");
+                            }
+
+                            // Use the producer client to send the batch of events to the event hub
+                            await _eventHubProducerClient.SendAsync(eventBatch, cancellationToken);
```

Comment:
Let's add a log here that reports the size of the EventDataBatch and the count of messages in the batch, so we can evaluate the average size.

Reply:
I actually created a new feature to enhance logging and metrics, so this can wait until later.